libs/java/service/src/main/java/ca/bc/gov/tno/services/data/TnoApi.java

```diff
@@ -0,0 +1,228 @@
+    this.config = config;
+    this.clientFactory = clientFactory;
+    this.client = createRestTemplate();
+  }
+
+  /**
+   * @return The current access token.
+   */
+  public TokenResponse getToken() {
+    return this.token;
+  }
+
+  private RestTemplate createRestTemplate() {
+    var xmlOrJson = new MappingJackson2HttpMessageConverter();
+    xmlOrJson.setSupportedMediaTypes(Arrays
+        .asList(new MediaType[] { MediaType.TEXT_XML, MediaType.APPLICATION_JSON, MediaType.APPLICATION_NDJSON }));
+    var form = new FormHttpMessageConverter();
+    return createRestTemplate(Arrays.asList(new HttpMessageConverter<?>[] { xmlOrJson, form }));
+  }
+
+  private RestTemplate createRestTemplate(List<HttpMessageConverter<?>> messageConverters) {
+    var rest = new RestTemplate(messageConverters);
+    return rest;
+  }
+
+  /**
+   * Make AJAX request to keycloak to fetch access token.
+   * 
+   * @return The token response.
+   * @throws ApiException If the AJAX request fails for some reason.
+   */
+  public TokenResponse requestToken() throws ApiException {
+    var url = config.getAuthorityUrl() + "/realms/" + config.getRealm() + "/protocol/openid-connect/token";
+    var headers = new HttpHeaders();
+    headers.setAccept(Arrays.asList(MediaType.APPLICATION_JSON));
+    headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+    var map = new LinkedMultiValueMap<String, String>();
+    map.add("audience", config.getClientId());
+    map.add("grant_type", "client_credentials");
+    map.add("client_id", config.getClientId());
+    map.add("client_secret", config.getClientSecret());
+    var entity = new HttpEntity<MultiValueMap<String, String>>(map, headers);
+    var response = client.exchange(url, HttpMethod.POST, entity, TokenResponse.class);
+
+    var status = response.getStatusCode();
+    if (status.series() == HttpStatus.Series.SUCCESSFUL) {
+      this.token = response.getBody();
+      return this.token;
+    } else {
+      throw new ApiException(String.format("Failed to retrieve token: %s", status.getReasonPhrase()));
+    }
+  }
+
+  /**
+   * Make an AJAX request to the API.
+   * 
+   * @param <T>          The type of object.
+   * @param url          The URL to the endpoint.
+   * @param method       The HTTP method.
+   * @param data         The request body data.
+   * @param responseType The response type.
+   * @return A new instance of an object of the specified type.
+   * @throws ApiException            If an error occurred when communicating with
+   *                                 the API.
+   * @throws ResourceAccessException An error occurred while making an AJAX
+   *                                 request.
+   */
+  public <T> T request(String url, HttpMethod method, T data, Class<T> responseType)
+      throws ApiException, ResourceAccessException {
+    var token = requestToken();
+    var jsonClient = new JsonRestTemplate();
```

Comment:
Have to use custom `JsonRestTemplate` so that we can get the actual errors.  The standard `RestTemplate` hides the deserialization errors and throws an unrelated exception... Unbelievable that this is used by enterprises around the world.  What a joke!